modules/Users/entryPointSaveGoogleApiKey.php

```diff
@@ -0,0 +1,61 @@
+<?php
+if (!defined('sugarEntry') || !sugarEntry) {
+    die('Not A Valid Entry Point');
+}
+
+global $sugar_config;
+
+$client = new Google_Client();
+$client->setApplicationName('SuiteCRM');
+$client->setScopes(Google_Service_Calendar::CALENDAR);
+$json = base64_decode($sugar_config['google_auth_json']);
+if (!$config = json_decode($json, true)) {
+    die('Invalid json for auth config');
+}
+$client->setAuthConfig($config);
+$client->setAccessType('offline');
+
+if (isset($_REQUEST['getnew'])) {
+    $authUrl = $client->createAuthUrl();
+    SugarApplication::redirect($authUrl);
+}
+
+if (isset($_REQUEST['code'])) {
+    global $GLOBALS;
+    global $current_user;
+    $user= new user();
+    $user->retrieve($current_user->id);
+    $accessToken = $client->fetchAccessTokenWithAuthCode($_REQUEST['code']);
+    $user->setPreference('GoogleApiToken', base64_encode(json_encode($accessToken)), false, 'GoogleSync');
```

Comment:
$user->retrieve() could returne null and it could be indicates an error. please handle it.